google_guest_agent/hostname/hostname.go

```diff
@@ -0,0 +1,241 @@
+	"path"
+	"runtime"
+	"strings"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/cfg"
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/command"
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/events"
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/events/metadata"
+	mds "github.com/GoogleCloudPlatform/guest-agent/metadata"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+var (
+	// ReconfigureHostnameCommand is the command id registered for hostname configuration
+	ReconfigureHostnameCommand = "agent.hostname.reconfigurehostname"
+	disallowedConfigurations   = map[string]bool{"": true, "metadata.google.internal": true}
+	hostnameFqdnMu             = new(sync.Mutex)
+	lastHostname               string //As retrieved from MDS
+	lastFqdn                   string //As retrieved from MDS
+)
+
+// ReconfigureHostnameRequest is the structure of requests to the ReconfigureHostnameCommand
+// Regardless of request options, guest agent will only set the hostname or fqdn if enabled in the MDS.
+type ReconfigureHostnameRequest struct {
+	command.Request
+}
+
+// ReconfigureHostnameResponse is the structure of responses from the ReconfigureHostnameCommand
+// Status code meanings:
+// 0: everything ok
+// 1: error setting hostname
+// 2: error setting fqdn
+// 3: error setting hostname and fqdn
+type ReconfigureHostnameResponse struct {
+	command.Response
+	// Hostname is the hostname which was set. Empty if unset, either due to configuration or error.
+	Hostname string
+	// Fqdn is the hostname which was set. Empty if unset, either due to configuration or error.
+	Fqdn string
+}
+
+// Init registers a hostname command handler and subscribes to the metadata longpoll event if the user has enabled network interface setup and either hostname or fqdn management.
+func Init(ctx context.Context, mdsclient mds.MDSClientInterface) {
+	if cfg.Get().Unstable.SetFqdn || cfg.Get().Unstable.SetHostname {
+		fqdn, err := mdsclient.GetKey(ctx, path.Join("instance", "hostname"), nil)
+		if err != nil {
+			logger.Errorf("could not get metadata hostname from MDS: %v", err)
+		} else if cfg.Get().Unstable.FqdnAsHostname {
+			lastHostname = fqdn
+			lastFqdn = fqdn
+		} else {
+			hostname, _, ok := strings.Cut(fqdn, ".")
+			if !ok {
+				logger.Errorf("metadata hostname %s is not an fqdn", fqdn)
+			} else {
+				lastFqdn = fqdn
+				lastHostname = hostname
+			}
+			b, err := ReconfigureHostname(nil)
+			if err != nil {
+				logger.Errorf("failed to call reconfigurehostname during setup: %v", err)
+			} else {
+				var resp ReconfigureHostnameResponse
+				err := json.Unmarshal(b, &resp)
+				if err != nil {
+					logger.Errorf("malformed response from reconfigurehostname: %v", err)
+				} else if resp.Status != 0 {
+					logger.Errorf("error reconfiguring hostname: %s", resp.StatusMessage)
+				}
+			}
+		}
+		events.Get().Subscribe(metadata.LongpollEvent, nil, processMdsEvent)
```

Comment:
we are querying hostname in the Init but kind of doing the same in response to event, wouldn't that be worth having that only on the event handler?

Reply:
Unless I misunderstood the API, this event won't fire immediately but we still want to configure the hostname right now.

But actually I reread the documentation for the platform feature for the hostname today and it is not possible to change this metadata key after instance creation. If you have no objections I'd rather just drop the metadata event signal since the `instance/hostname` key will never change, and only do the initial setup.